File: go/day3.go
package main

import (
	"fmt"
	"regexp"
	"strconv"
)

func Day3a() {
	lines, err := ReadLines("../inputs/day3.txt")
	if err != nil {
		panic("Could not read lines")
	}

	str := ""
	for _, line := range lines {
		str += line
	}

	pattern := `mul\((\d{1,3}),(\d{1,3})\)`
	re := regexp.MustCompile(pattern)
	matches := re.FindAllStringSubmatch(str, -1)

	sum := 0
	for _, m := range matches {
		num1, _ := strconv.Atoi(m[1])
		num2, _ := strconv.Atoi(m[2])
		sum += num1 * num2
	}
	fmt.Printf("%v\n", sum)
}

func Day3b() {
	lines, err := ReadLines("../inputs/day3.txt")
	if err != nil {
		panic("Could not read lines")
	}

	str := ""
	for _, line := range lines {
		str += line
	}

	pattern := `(mul\((\d{1,3}),(\d{1,3})\))|(do\(\))|(don't\(\))`
	re := regexp.MustCompile(pattern)
	matches := re.FindAllStringSubmatch(str, -1)

	sum := 0
	add := true
	for _, m := range matches {
		if m[4] != "" {
			add = true
			continue
		}
		if m[5] != "" {
			add = false
			continue
		}
		if add {
			num1, _ := strconv.Atoi(m[2])
			num2, _ := strconv.Atoi(m[3])
			sum += num1 * num2
		}
	}
	fmt.Printf("%v\n", sum)
}

File: python/day3.py
import re


def day3a() -> None:
    string = ""
    with open("../inputs/day3.txt", "r") as file:
        for line in file:
            string += line

    pattern = r"mul\((\d{1,3}),(\d{1,3})\)"
    results = re.findall(pattern, string)

    sum_product = 0
    for res in results:
        sum_product += int(res[0]) * int(res[1])
        # print(type(res))
    print(sum_product)


def day3b() -> None:
    string = ""
    with open("../inputs/day3.txt", "r") as file:
        for line in file:
            string += line

    pattern = r"(mul\((\d{1,3}),(\d{1,3})\))|(do\(\))|(don't\(\))"
    results = re.findall(pattern, string)

    add = True
    sum_product = 0
    for res in results:
        if res[3] != "":
            add = True
            continue
        if res[4] != "":
            add = False
            continue

        if add:
            sum_product += int(res[1]) * int(res[2])

    print(sum_product)


day3b()
